docs/proposals/ssh-known_hosts.rst

```diff
@@ -0,0 +1,93 @@
+.. vim:set ft=rst spell:
+
+Managing SSH Known Hosts
+========================
+
+This proposal suggests a method to more securely manage and ensure
+authentication of remote hosts through the SSH protocol within Tarmak
+environments on AWS.
+
+Background
+----------
+
+Currently we solely use the external OpenSHH SSH command line tool to connect to
+remote instances on EC2 for both interactive shell sessions as well as
+tunnelling and proxy commands for other services, including, connecting to the
+Vault cluster and the private Kubernetes API server endpoint. Currently in
+development is the replacement of our programmatic use cases of SSH in favour of
+the in package Go solution, a choice stemming from pain points developing more
+sophisticated utility functions for Tarmak and the desire for improvements in
+control of connections to remote hosts.
+
+During development of this replacement it became clear that proper care must be
+taken during authentication of host public keys during connection and manual
+management of our ``ssh_known_hosts`` cluster file. Our current implementation
+allows OpenSSH to maintain this file however, does not exit with an error if
+public keys do not match due to the flag ``StrictHostKeyChecking`` set to
+``no``. Not only does a miss-match in public keys not cause an error, the
+population of known public keys on different authenticated machines to the same
+EC2 hosts will always use the hosts presented public key, meaning the set of
+public keys could potentially be different for users accessing the same cluster.
+
+Objective
+---------
+
+By implementing stricter enforcement of the ``ssh_known_hosts`` file and passing
+it's management to Tarmak, we can improve the security of SSH connections to
+remote hosts. The key high level points to achieving this is as follows:
+
+ - Disable writes from the OpenSSH command to the ``ssh_known_hosts`` file and
```

Comment:
Would it be easier for tarmak to maintain an 'in memory' ssh_known_hosts map (that could be 'generated' when tarmak needs to SSH to instances), and then add a subcommand `tarmak get known-hosts` or something in case users want to access/get a copy of the file?

Perhaps there's some other reason why you'd rather share a single file for this that I've missed! 😄

Reply:
I quite like having this file around as it makes it easier for using ssh without tarmak - we also keep the SSH config around. Same idea as having the kubeconfig in file to use kubectl without tarmak.

Reply:
Sure thing - I just thought it might be simpler as you won't have to deal with user-introduced unknowns!